language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines the (acyclic) borrow graph for the type and memory safety analysis.
+use crate::nonce::Nonce;
+use mirai_annotations::{checked_postcondition, checked_precondition};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// There are two types of edges in the borrow graph.
+/// A strong edge carries more information than a weak edge.
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum EdgeType {
+    Strong,
+    Weak,
+}
+
+/// The label on an edge
+pub type Label<T> = Vec<T>;
+
+/// A labeled edge in borrow graph
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Edge<T> {
+    edge_type: EdgeType,
+    label: Label<T>,
+    to: Nonce,
+}
+
+/// A borrow graph is represented as a map from a source nonce to the set of all edges
+/// coming out of it.
+#[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+pub struct BorrowGraph<T>(BTreeMap<Nonce, BTreeSet<Edge<T>>>);
+
+impl<T> BorrowGraph<T>
+where
+    T: Ord,
+    T: Clone,
+{
+    /// creates a new empty borrow graph
+    pub fn new() -> Self {
+        BorrowGraph(BTreeMap::new())
+    }
+
+    /// adds a fresh nonce
+    pub fn add_nonce(&mut self, nonce: Nonce) {
+        checked_precondition!(!self.0.contains_key(&nonce));
+        self.0.insert(nonce, BTreeSet::new());
+    }
+
+    /// adds a weak edge
+    pub fn add_weak_edge(&mut self, from: Nonce, label: Label<T>, to: Nonce) {
+        checked_precondition!(self.0.contains_key(&from));
```

Comment:
One of the advantages of using an actual field name is that you'll have to decide whether or it it should be public. If not, I get to tell you it is a bad idea to use private state in a precondition. :-)

Reply:
You have to decide if its public or not even in the current set up. Rust doesn't really have a concept of "private" in the traditional OO sense, its more of an "internal" to the module visibility modifier. 

If the field was public, it would be
```
pub struct BorrowGraph<T>(pub BTreeMap<Nonce, BTreeSet<Edge<T>>>);
```